### ISSUE TITLE
Explicit padding mode

### ISSUE THREAD

**Issue by rwightman** [google/objax#30]
It looks like objax currently limits padding to one of VALID or SAME. This prevents the ability to use explicit padding and would prevent compatibility with models from PyTorch, Gluon that only support explicit (symmetric) padding without adding extra Pad layers to the model.

It'd be nice to at minimum add the ability to support TF style explicit padding (specify both sides of every dim), the underlying jax conv impl is able to receive a `[[0, 0], [pad_beg, pad_end],[pad_beg, pad_end], [0, 0]]` spec like other low level TF conv.

Even nicer would be a simplificed, per-spatial dim symmetric values like PyTorch, Gluon `[pad_h, pad_w]` or just `pad` . My default for most 2D convnets in PyTorch is to use `pad = ((stride - 1) + dilation * (kernel_size - 1)) // 2`, which results in a 'same-ish' padding value. This can always be done on top of the full low/high padding sequence above. 

Some TF models explicitly work around the limitations of SAME padding. By limitations, I mean the fact that you end up with input dependent padding that can be aysmmetric and shift your feature maps relative to each other in a manner that varies as you change your input size.
https://github.com/tensorflow/models/blob/146a37c6663e4a249e02d3dff0087b576e3dc3a1/research/deeplab/core/xception.py#L81-L201

Possible interfaces:

* `padding : Union[ConvPadding, Sequence[Tuple[int, int]]]` (like `conv_general_dilated` but with the enum for valid/same)

* Add more modes the enum and associated values for those that need it via a dataclass

```
class PaddingType(enum.Enum):
  """An Enum holding the possible padding values for convolution modules."""
    SAME = 'SAME'
    VALID = 'VALID'
    RAW = 'RAW'  # specify padding as seq of high/low tuples
    SYM = 'SYM'  # specify symmetric padding for spatial dim as tuple for H, W or single int

@dataclass
class Padding:
    type: PaddingType = PaddingType.SAME
    value: Union[Sequence[Tuple[int, int]], Tuple[int, int], int] = None

    @classmethod
    def same(cls):
        return Padding(PaddingType.SAME)

    @classmethod
    def valid(cls):
        return Padding(PaddingType.VALID)

    @classmethod
    def raw(cls, value: Sequence[Tuple[int, int]]):
        return Padding(PaddingType.RAW, value=value)

    @classmethod
    def sym(cls, value: Union[Tuple[int, int], int]):
        return Padding(PaddingType.SYM, value=value)
```

**Comment by david-berthelot** [comment]
I like your proposal. If I understand correctly, in this new way of doing things a call to `tf.pad` would be added inside the code for `Conv2D` to handle more padding modes (like symmetric, reflect, constant padding)? And when not needed (zero-padding or valid) there would be no call to `tf.pad` (to keep performance optimized)? And the API change would also permits passing explicit padding intervals (integers) per axis.

So the convolution padding would be defined as (to define the "geometry" of the padding):
```python
class ConvPadding(enum.Enum):
  """An Enum holding the possible padding values for convolution modules."""
    SAME = 'SAME'
    VALID = 'VALID'
```

And the mode (what data do we put inside the padding):
```python
class ConvPaddingMode(enum.Enum):
  """An Enum holding the possible padding values for convolution modules."""
    SYM = 'SYM'  # specify symmetric padding for spatial dim as tuple for H, W or single int
    REFLECT = 'REFLECT'
    MEAN = 'MEAN'
    # ... and so on.
```

If I understand correctly, that's what you encapsulated in the proposed `Padding` dataclass.
I would prefer to pass two arguments to __call__ for Conv2D instead (to keep the code accessible, we voluntarily limit abstractions to keep the code more welcoming to people of varied CS experience).
```python
ConvPaddingInt = Union[Sequence[Tuple[int, int]], Tuple[int, int], int]

class Conv2D(...):
    def __init__(self,
                 ...
                 padding: Union[ConvPadding, ConvPaddingInt]  = ConvPadding.SAME,
                 padding_mode: Union[ConvPaddingMode, float] = 0,
                 ...
```

Would this variant  of your initial proposal work for you?

**Comment by rwightman** [comment]
@david-berthelot TLDR yes, I think the above proposal would work great despite confusion over my poor choice of the term 'symmetric'

Regarding 'symmetric', I wasn't actually thinking modes while trying to jot down the idea. My additional enum values were all ways of specifying the 'geometries' (or 'padding type'). I was proposing different alternatives dataclass w/ extended enum vs union as I wasn't sure if single types were preferred over unions for interfaces here.

So 'symmetric' referred to the fact that the int and Tuple[int, int] result in symmetric amounts of padding -- pad_beg == pad_end with no ability to end up with beg != end as with 'SAME' padding.

As for the modes, that'd also be useful for some less common use cases and covered by calling pad before the conv as you mention. I use 'circular' and 'constant' from time to time, but not 'reflect', 'symmetric', 'replicate'.

**Comment by david-berthelot** [comment]
Would you be open to contributing a PR? If not please let me know; I, or someone else, can follow up.

**Comment by rwightman** [comment]
@david-berthelot I would be, currently I have part of the change (padding type, not the mode) intertwined with some others for my prototyping. Once I get a bit further on that and get a chance to test properly I can do a PR here. But if someone else wants to take it on here before that happens, that's a-ok.

**Comment by david-berthelot** [comment]
Here's the plan, I collected more opinions and we'll extend in the simplest way, I'll do the coding in the coming days.

This part will remain the same.
```
class ConvPadding(enum.Enum):
  """An Enum holding the possible padding values for convolution modules."""
    SAME = 'SAME'
    VALID = 'VALID'
```

Conv2D (and others) will be extended to support numeric padding.
```
ConvPaddingInt = Union[Sequence[Tuple[int, int]], Tuple[int, int], int]

class Conv2D(...):
    def __init__(self,
                 ...
                 padding: Union[ConvPadding, ConvPaddingInt] = ConvPadding.SAME,
                 ...
```

In the end, we decided againt added numpy-style mode paddings since we want to be mindful about keeping the original design simplicity.

**Comment by rwightman** [comment]
@david-berthelot makes a lot of sense, the numeric padding was the motivation for this issue. Fancy modes are much less common and can always be done with separate padding layers.
